File: cpp-solutions/4.cpp
class Solution
{
  public:
    double findMedianSortedArrays(vector<int> &nums1, vector<int> &nums2)
    {
        int m = nums1.size(), n = nums2.size();
        int cnt = 0, i = 0, j = 0;
        double res = 0.0;
        bool odd = (n + m) % 2 != 0;
        int tar = odd ? (n + m) / 2 + 1 : (n + m) / 2;
        while (i < m || j < n)
        {
            int t;
            if (j >= n || i < m && j < n && nums1[i] <= nums2[j])
            {
                t = nums1[i++];
            }
            else if (i >= m || i < m && j < n && nums1[i] > nums2[j])
            {
                t = nums2[j++];
            }
            cnt++;
            if (cnt == tar)
            {
                res += t;
                if (odd)
                    break;
            }
            if (cnt == tar + 1)
            {
                res = (res + t) / 2;
                break;
            }
        }
        return res;
    }
};

File: golang/4/4.go
package main

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	var (
		odd        bool
		count, mid int
		res        float64
		len1       = len(nums1)
		len2       = len(nums2)
	)

	if (len1+len2)%2 != 0 {
		odd = true
		mid = (len1+len2)/2 + 1
	} else {
		mid = (len1 + len2) / 2
	}

	for i, j := 0, 0; i < len1 || j < len2; {
		t := 0
		if j >= len2 || (i < len1 && j < len2 && nums1[i] <= nums2[j]) {
			t = nums1[i]
			i++
		} else if i >= len1 || (i < len1 && j < len2 && nums1[i] > nums2[j]) {
			t = nums2[j]
			j++
		}

		count++
		switch count {
		case mid:
			res += float64(t)
			if odd {
				return res
			}
		case mid + 1:
			res += float64(t)
			return res / 2
		}
	}

	return res
}
